lib/view_model/active_record/update_data.rb

```diff
@@ -90,71 +366,86 @@ module Schemas
-            # constraint in code.
-          },
-        }
-      )
-
-      REMOVE_ACTION = JsonSchema.parse!(remove)
-
-      collection_update =
+      append_mixin = {
+        'description'          => 'collection append',
+        'additionalProperties' => false,
+        'properties'           => {
+          ViewModel::TYPE_ATTRIBUTE => { 'enum' => [FunctionalUpdate::Append::NAME] },
+          BEFORE_ATTRIBUTE          => viewmodel_reference_only,
+          AFTER_ATTRIBUTE           => viewmodel_reference_only
+        },
+      }
+
+      fupdate_owned =
+        fupdate_base.(ViewModel::Schemas::VIEWMODEL_UPDATE_SCHEMA)
+
+      fupdate_shared           =
+        fupdate_base.({ 'oneOf' => [ViewModel::Schemas::VIEWMODEL_REFERENCE_SCHEMA,
+                                    viewmodel_reference_only] })
+
+      # Referenced updates are special:
+      #  - Append requires `_ref` hashes
+      #  - Update requires `_ref` hashes
+      #  - Remove requires vm refs (type/id)
+      # Checked in code (ReferencedCollectionUpdate::Builder.parse_*_values)
+
+      APPEND_ACTION            = JsonSchema.parse!(fupdate_owned.deep_merge(append_mixin))
+      REFERENCED_APPEND_ACTION = JsonSchema.parse!(fupdate_shared.deep_merge(append_mixin))
+
+      update_mixin = {
+        'description'          => 'collection update',
+        'additionalProperties' => false,
+        'properties'           => {
+          ViewModel::TYPE_ATTRIBUTE => { 'enum' => [FunctionalUpdate::Update::NAME] }
+        },
+      }
+
+      UPDATE_ACTION            = JsonSchema.parse!(fupdate_owned.deep_merge(update_mixin))
+      REFERENCED_UPDATE_ACTION = JsonSchema.parse!(fupdate_shared.deep_merge(update_mixin))
+
+      remove_mixin = {
+        'description'          => 'collection remove',
+        'additionalProperties' => false,
+        'properties'           => {
+          ViewModel::TYPE_ATTRIBUTE => { 'enum' => [FunctionalUpdate::Remove::NAME] },
+          # The VALUES_ATTRIBUTE should be a viewmodel_reference, but in the
+          # name of error messages, we allow more keys and check the
+          # constraint in code.
+        },
+      }
+
+      REMOVE_ACTION            = JsonSchema.parse!(fupdate_owned.deep_merge(remove_mixin))
+      REFERENCED_REMOVE_ACTION = JsonSchema.parse!(fupdate_shared.deep_merge(remove_mixin))
+
+
+      collection_update = ->(base_schema) do
         {
           'type'                 => 'object',
           'description'          => 'collection functional update',
           'additionalProperties' => false,
           'required'             => [ViewModel::TYPE_ATTRIBUTE, ACTIONS_ATTRIBUTE],
           'properties'           => {
             ViewModel::TYPE_ATTRIBUTE => { 'enum' => [FUNCTIONAL_UPDATE_TYPE] },
-            ACTIONS_ATTRIBUTE         => { 'type' => 'array', 'items' => base_functional_update_schema }
+            ACTIONS_ATTRIBUTE => { 'type' => 'array', 'items' => base_schema }
```

Comment:
might still want to check at this level that the TYPE is `{ 'enum' => [update, append, remove]}`, and then as you say, leave the specific validation of each of those to the subschemas.